File: Java/src/No1763.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No1763 {
    public String longestNiceSubstring(String s) {
        int n = s.length();
        String ans = "";
        char[] cs = s.toCharArray();
        for (int i = 0; i < n; ++i) {
            int a = 0, b = 0;
            for (int j = i; j < n; ++j) {
                if (Character.isLowerCase(cs[j])) a |= 1 << (cs[j] - 'a');
                else b |= 1 << (cs[j] - 'A');
                if (a == b && j - i + 1 > ans.length()) ans = s.substring(i, j + 1);
            }
        }
        return ans;
    }
}

File: CPP/No1763.cc
/**
 * Created by Xiaozhong on 2022/2/1.
 * Copyright (c) 2022/2/1 Xiaozhong. All rights reserved.
 */
#include "iostream"
#include "vector"
#include "string"
#include "algorithm"
#include "unordered_map"
#include "unordered_set"

using namespace std;

class Solution {
public:
    string longestNiceSubstring(string s) {
        if (s.length() <= 1) return "";
        string ans;
        for (int start = 0; start < s.length(); ++start) {
            int lower = 0;
            int upper = 0;
            for (int end = start; end < s.length(); ++end) {
                if (islower(s[end])) lower |= 1 << (s[end] - 'a');
                else if (isupper(s[end])) upper |= 1 << (s[end] - 'A');
                if ((lower ^ upper) == 0 && end - start + 1 > ans.length()) ans = s.substr(start, end - start + 1);
            }
        }
        return ans;
    }
};

int main() {
    Solution s;
    cout << s.longestNiceSubstring("YazaAay");
}